# Heart Disease Prediction — Dataset Validation

## UCI Heart Disease Dataset — Cleveland Subset

### Purpose

This notebook validates the raw Cleveland subset of the UCI Heart Disease dataset before preprocessing and machine learning.

The validation includes:

- Dataset dimensions
- Column names and data types
- Missing-value detection
- Duplicate-row detection
- Original target distribution
- Unique-value validation
- Binary target construction

The raw dataset is not modified in this notebook.


In [1]:
from pathlib import Path
import pandas as pd

PROJECT_ROOT = Path.cwd().parent if Path.cwd().name == "notebooks" else Path.cwd()
if str(PROJECT_ROOT) not in __import__("sys").path:
    __import__("sys").path.insert(0, str(PROJECT_ROOT))

from src.preprocessing import load_data, validate_columns, prepare_features_target

DATA_PATH = PROJECT_ROOT / "data" / "raw" / "processed.cleveland.data"
print("Dataset path:", DATA_PATH.resolve())
print("Dataset exists:", DATA_PATH.exists())

df = load_data(DATA_PATH)
validate_columns(df, require_target=True)
print("Dataset loaded successfully using src.preprocessing.load_data().")


Dataset path: C:\Projects\heart-disease-prediction\data\raw\processed.cleveland.data
Dataset exists: True
Dataset loaded successfully using src.preprocessing.load_data().


In [2]:
print("Shape:", df.shape)

print("\nColumn names:")
print(df.columns.tolist())


Shape: (303, 14)

Column names:
['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach', 'exang', 'oldpeak', 'slope', 'ca', 'thal', 'num']


In [3]:
print("First 5 records:")
display(df.head())

print("\nLast 5 records:")
display(df.tail())


First 5 records:


,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,num
0,63.0,1.0,1.0,145.0,233.0,1.0,2.0,150.0,0.0,2.3,3.0,0.0,6.0,0
1,67.0,1.0,4.0,160.0,286.0,0.0,2.0,108.0,1.0,1.5,2.0,3.0,3.0,2
2,67.0,1.0,4.0,120.0,229.0,0.0,2.0,129.0,1.0,2.6,2.0,2.0,7.0,1
3,37.0,1.0,3.0,130.0,250.0,0.0,0.0,187.0,0.0,3.5,3.0,0.0,3.0,0
4,41.0,0.0,2.0,130.0,204.0,0.0,2.0,172.0,0.0,1.4,1.0,0.0,3.0,0



Last 5 records:


,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,num
298,45.0,1.0,1.0,110.0,264.0,0.0,0.0,132.0,0.0,1.2,2.0,0.0,7.0,1
299,68.0,1.0,4.0,144.0,193.0,1.0,0.0,141.0,0.0,3.4,2.0,2.0,7.0,2
300,57.0,1.0,4.0,130.0,131.0,0.0,0.0,115.0,1.0,1.2,2.0,1.0,7.0,3
301,57.0,0.0,2.0,130.0,236.0,0.0,2.0,174.0,0.0,0.0,2.0,1.0,3.0,1
302,38.0,1.0,3.0,138.0,175.0,0.0,0.0,173.0,0.0,0.0,1.0,NaN,3.0,0


In [4]:
print("Data types:")
print(df.dtypes)


Data types:
age         float64
sex         float64
cp          float64
trestbps    float64
chol        float64
fbs         float64
restecg     float64
thalach     float64
exang       float64
oldpeak     float64
slope       float64
ca          float64
thal        float64
num           int64
dtype: object


In [5]:
missing_counts = df.isna().sum()

print("=== Missing-Value Report ===")
print(missing_counts[missing_counts > 0])

print("\n=== Total Missing Values ===")
print(df.isna().sum().sum())

missing_rows = df[df.isna().any(axis=1)]

print("\n=== Rows Containing Missing Values ===")
display(missing_rows)


=== Missing-Value Report ===
ca      4
thal    2
dtype: int64

=== Total Missing Values ===
6

=== Rows Containing Missing Values ===


,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,num
87,53.0,0.0,3.0,128.0,216.0,0.0,2.0,115.0,0.0,0.0,1.0,0.0,NaN,0
166,52.0,1.0,3.0,138.0,223.0,0.0,0.0,169.0,0.0,0.0,1.0,NaN,3.0,0
192,43.0,1.0,4.0,132.0,247.0,1.0,2.0,143.0,1.0,0.1,2.0,NaN,7.0,1
266,52.0,1.0,4.0,128.0,204.0,1.0,0.0,156.0,1.0,1.0,2.0,0.0,NaN,2
287,58.0,1.0,2.0,125.0,220.0,0.0,0.0,144.0,0.0,0.4,2.0,NaN,7.0,0
302,38.0,1.0,3.0,138.0,175.0,0.0,0.0,173.0,0.0,0.0,1.0,NaN,3.0,0


In [6]:
duplicate_count = df.duplicated().sum()

print("Number of duplicate rows:", duplicate_count)

if duplicate_count > 0:
    print("\nDuplicate rows:")
    display(df[df.duplicated(keep=False)])
else:
    print("No duplicate rows found.")


Number of duplicate rows: 0
No duplicate rows found.


In [7]:
target_counts = df["num"].value_counts().sort_index()

print("=== Original Target Distribution ===")
print(target_counts)

print("\n=== Original Target Percentages ===")
print(
    df["num"]
    .value_counts(normalize=True)
    .sort_index()
    .mul(100)
    .round(2)
)

=== Original Target Distribution ===
num
0    164
1     55
2     36
3     35
4     13
Name: count, dtype: int64

=== Original Target Percentages ===
num
0    54.13
1    18.15
2    11.88
3    11.55
4     4.29
Name: proportion, dtype: float64


In [8]:
print("=== Unique Values ===")

categorical_check = [
    "sex",
    "cp",
    "fbs",
    "restecg",
    "exang",
    "slope",
    "ca",
    "thal",
    "num"
]

for column in categorical_check:
    print(f"{column}: {sorted(df[column].dropna().unique())}")


=== Unique Values ===
sex: [np.float64(0.0), np.float64(1.0)]
cp: [np.float64(1.0), np.float64(2.0), np.float64(3.0), np.float64(4.0)]
fbs: [np.float64(0.0), np.float64(1.0)]
restecg: [np.float64(0.0), np.float64(1.0), np.float64(2.0)]
exang: [np.float64(0.0), np.float64(1.0)]
slope: [np.float64(1.0), np.float64(2.0), np.float64(3.0)]
ca: [np.float64(0.0), np.float64(1.0), np.float64(2.0), np.float64(3.0)]
thal: [np.float64(3.0), np.float64(6.0), np.float64(7.0)]
num: [np.int64(0), np.int64(1), np.int64(2), np.int64(3), np.int64(4)]


## Binary Target Definition

The original UCI diagnosis variable (`num`) contains five values:

- `0` — absence of heart disease
- `1–4` — presence of heart disease at different diagnostic levels

For this study, the prediction problem is formulated as binary classification.

The original `num` variable is retained for reference, while a new `target` variable is created:

- `target = 0` → No heart disease
- `target = 1` → Heart disease

In [9]:
X, y = prepare_features_target(df)
df["target"] = y

print("=== Binary Target Distribution ===")
print(df["target"].value_counts().sort_index())

print("\n=== Binary Target Percentages ===")
print(
    df["target"]
    .value_counts(normalize=True)
    .sort_index()
    .mul(100)
    .round(2)
)


=== Binary Target Distribution ===
target
0    164
1    139
Name: count, dtype: int64

=== Binary Target Percentages ===
target
0    54.13
1    45.87
Name: proportion, dtype: float64


In [10]:
print("========================================")
print("FINAL DATASET VALIDATION SUMMARY")
print("========================================")

X, y = prepare_features_target(df)

print(f"Records: {df.shape[0]}")
print("Predictor variables: 13")
print("Original target variable: num")
print("Derived target variable: target")
print(f"Total columns currently in DataFrame: {df.shape[1]}")

print(f"\nTotal missing values: {df.isna().sum().sum()}")
print(f"Number of duplicate rows: {df.duplicated().sum()}")

print("\nOriginal target:")
print(df["num"].value_counts().sort_index())

print("\nBinary target:")
print(y.value_counts().sort_index())


FINAL DATASET VALIDATION SUMMARY
Records: 303
Predictor variables: 13
Original target variable: num
Derived target variable: target
Total columns currently in DataFrame: 15

Total missing values: 6
Number of duplicate rows: 0

Original target:
num
0    164
1     55
2     36
3     35
4     13
Name: count, dtype: int64

Binary target:
target
0    164
1    139
Name: count, dtype: int64


### Duplicate Records

An exact duplicate-row check was performed on the 303 observations. No duplicate rows were identified.

Therefore, no duplicate records were removed from the dataset.


## Conclusion

The raw Cleveland subset of the UCI Heart Disease dataset was successfully loaded and validated.

The dataset contains 303 patient records and 13 predictor variables with the original diagnostic variable `num`. Six missing values were identified: four in `ca` and two in `thal`. The original target contains five diagnostic categories (0–4).

For the proposed binary classification problem, a derived `target` variable was created, where 0 represents absence of heart disease and 1 represents presence of heart disease.

No data imputation, duplicate removal, scaling, encoding, or model training is performed in this notebook. These operations will be addressed in subsequent preprocessing and machine-learning stages.